composer/utils/object_store/gs_object_store.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Google Cloud SDK - Compatible object store."""
+
+from __future__ import annotations
+
+import os
+import pathlib
+import uuid
+from typing import Callable, Optional, Union
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore
+
+__all__ = ['GsObjectStore']
+
+BOTOCORE_CLIENT_ERROR_CODES = ('403', '404', 'NoSuchKey')
+
+
+def _reraise_gs_errors(uri: str, e: Exception):
+    try:
+        from google.api_core.exceptions import GatewayTimeout, NotFound
+
+        print('Reraising exception: {e.message}')
+
+        # If it's a google service NotFound error
+        if isinstance(e, NotFound):
+            raise FileNotFoundError(f'Object {uri} not found.') from e
+
+        # All clienterror (HTTP 4xx) responses
+        elif isinstance(e, GatewayTimeout):
+            raise ValueError(f'Time out when uploading/downloading {uri} using google cloud storage') from e
+
+    except ImportError as e:
+        raise MissingConditionalImportError(conda_package='google-cloud-storage',
+                                            extra_deps_group='google-cloud-storage',
+                                            conda_channel='conda-forge') from e
+    # Otherwise just raise the original error.
+    raise e
+
+
+class GsObjectStore(ObjectStore):
+    """Utility for uploading to and downloading from an Google Cloud bucket using :mod:`google cloud storage sdk` with either HMAC and service account authentications. If both authentiations are available, use service account authentication.
+
+    .. warning::
+
+        For best security practices, it is recommended to set credentials via environment variables or config files.
+
+        See :ref:`guide to credentials <boto3:guide_credentials>` for more information.
+
+    Args:
+        remote_dir (str, optional): Required. The URL to an Gooogle Cloud Storage object, formatted as scheme://bucket/path
+    """
+
+    def __init__(
+            self,
+            gs_root_dir: str  #  = 'gs://mosaicml-composer-tests/streaming/',
+    ) -> None:
+        try:
+            import boto3
+            from google.cloud.storage import Client
+        except ImportError as e:
+            raise MissingConditionalImportError('streaming', 'boto3') from e
+
+        if 'GOOGLE_APPLICATION_CREDENTIALS' in os.environ:
+            service_account_path = os.environ['GOOGLE_APPLICATION_CREDENTIALS']
+            self.client = Client.from_service_account_json(service_account_path)
+        elif 'GCS_KEY' in os.environ and 'GCS_SECRET' in os.environ:
+            # Create a session and use it to make our client. Unlike Resources and Sessions,
+            # clients are generally thread-safe.
+
+            session = boto3.session.Session()
```

Comment:
do we only need boto3 for this code path? if so, can we scope the import try/except to this code path?

Reply:
yup, thats what i meant, thanks!